rocr: GPU core file location support

cfreeamd opened the pull request:
## Motivation

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->
This pull request introduces new configuration options for controlling GPU core dump behavior and progress reporting via environment variables in the `Flag` class, along with minor comment corrections. The main focus is on enabling or disabling core dumps, customizing core dump patterns, and optionally showing progress during core dumps.

New environment variable support and configuration:

* Added support for the following environment variables in the `Flag` class to control core dump behavior and progress reporting: `HSA_COREDUMP_SHOW_PROGRESS`, `HSA_DISABLE_COREDUMP_ON_EXCEPTION`, and `HSA_COREDUMP_PATTERN`. These are parsed and stored as new member variables. [[1]](diffhunk://#diff-71e512729a5bfc7292d5c880bfc6fb648ac01e335093b9eb471aaa0b083bd979R301-R308) [[2]](diffhunk://#diff-71e512729a5bfc7292d5c880bfc6fb648ac01e335093b9eb471aaa0b083bd979R544-R547)
* Added corresponding accessor methods: `core_dump_disable()`, `enable_core_dump_progress()`, and `core_dump_pattern()` to expose these new configuration options.

Minor improvements:

* Fixed typos in comments related to the fallback core dump logic in both `amd_aql_queue.cpp` and `runtime.cpp`. [[1]](diffhunk://#diff-93eca645297abc8e958267f5cd1064c02d4f4a0a75016fb91f9de6b47163d607L1326-R1327) [[2]](diffhunk://#diff-48c26de4935a1e51070ee213f6c83eed9ff3ebd150c1e24effa1ee99c99d4f4dL2239-R2239)
## Technical Details


## Test Plan
https://github.com/ROCm/rocm-systems/pull/1871

## Test Result
Passed locally

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

cfreeamd committed:
rocr: WIP Support dump of GPU core file

cfreeamd committed:
WIP new core dump tests compile

cfreeamd committed:
WIP: anony namespaces, test updates, progress

Added disabled Fault test. Other non-disabled coredump tests don't work.

cfreeamd commented:
SWDEV-562769
The functionality in the ticket has been added, including the piping capability (Cline helped with this, so it went faster than anticipated). Let me know if I should disable the piping.

I have tested using the TestFaultExample, which is the usual TestExample, except we pass in nullptrs for the arrays. This is not meant to be run as part of CI, so it is disabled by default.  I tested the scenarios mentioned in the ticket.

Do not bother to review the other tests for now, they don't work. I will work on them now, but I wanted to allow you to start reviewing the implementation instead of waiting on getting those tests working.

lancesix commented on projects/rocr-runtime/runtime/hsa-runtime/core/util/flag.h:
```diff
@@ -298,6 +298,14 @@ class Flag {
     co_dmacopy_size_ = var.empty() ? 1024*1024 : atoi(var.c_str());
+
+    var = os::GetEnvVar("HSA_COREDUMP_SHOW_PROGRESS");
+    enable_core_dump_progress_ = (var == "1") ? true : false;
```
Pedantically, the expression `var == "1"` is a bool already, the ternary operator is redundant.

i.e.

```c++
     enable_core_dump_progress_ = var == "1";
```
is strictly equivalent.

cfreeamd replied:
ack

lancesix commented on projects/rocr-runtime/runtime/hsa-runtime/libamdhsacode/amd_core_dump.cpp:
```diff
@@ -72,6 +77,167 @@ namespace amd {
+#if defined(__linux__)
+  std::string result;
+  pid_t pid = getpid();
+  pid_t tid = syscall(SYS_gettid);
```
Why not use `gettid(2)`?

cfreeamd replied:
ack

palves commented on projects/rocr-runtime/runtime/hsa-runtime/libamdhsacode/amd_core_dump.cpp:
```diff
@@ -304,7 +470,222 @@ struct LoadSegmentBuilder : public SegmentBuilder {
+                                          bool show_progress) {
+#if defined(__linux__)
+  // Check if we're in a container
+  if (is_running_in_container()) {
```
ISTM that this unconditional container check prevents the "support pipe in HSA_COREDUMP_FILE inside container" case?  ISTM this check should be skipped if the 'pattern' came from the environment variable instead of core_pattern.

cfreeamd replied:
I missed that requirement. So, if the user is using HSA_COREDUMP_FILE with a pipe in a container we should assume the handler is accessible, unlike when  HSA_COREDUMP_FILE is not used?

palves replied:
Yes, I think that's what makes sense.

For the CPU core, a pipe program registered in core_pattern is always run by the host side (by the kernel), so it's a path visible from the host, and not from the container.

OTOH, the HSA_COREDUMP_FILE environment variable, like all env vars, is a userspace setting, independent of the kernel.  Thus, if someone sets up the HSA_COREDUMP_FILE var within a container, is because it points to a pipe program visible by the userspace running within the container, as that is all it can possible be.  The reason the runtime doesn't call the pipe program installed with core_pattern (because it's a host path) does not apply in this case.

lancesix commented on projects/rocr-runtime/runtime/hsa-runtime/core/util/flag.h:
```diff
@@ -298,6 +298,14 @@ class Flag {
+    var = os::GetEnvVar("HSA_DISABLE_COREDUMP_ON_EXCEPTION");
+    core_dump_disable_ = (var == "1") ? true : false;
+
+    core_dump_file_ = os::GetEnvVar("HSA_COREDUMP_FILE");
```
reading it, is seems it would be better name COREDUMP_PATTERN, this is not a proper filename at this point (and such naming would follow the kernel's kernel.core_pattern name).

cfreeamd replied:
ack

lancesix commented on projects/rocr-runtime/runtime/hsa-runtime/libamdhsacode/amd_core_dump.cpp:
```diff
@@ -72,6 +77,167 @@ namespace amd {
+    strncpy(hostname.data(), "unknown", hostname.size() - 1);
+    hostname[hostname.size() - 1] = '\0';
+  }
+  hostname[sizeof(hostname) - 1] = '\0';
```
2 things:
- Why use `hostname.size();` above and `sizeof(hostname)` here?
- The one in the if is redundant.

lancesix commented on projects/rocr-runtime/runtime/hsa-runtime/libamdhsacode/lnx/amd_core_dump.cpp:
```diff
@@ -441,30 +856,71 @@ hsa_status_t build_core_dump(const std::string& filename, const SegmentsInfo& se
+    if (show_progress) {
+      fprintf(stderr, "Generating GPU core dump via pipe handler\n");
+    }
+    return impl::write_to_pipe_handler(pattern, segments, rlimit.rlim_cur, show_progress);
```
I am not sure `rlimit.rlim_cur` should be used here. It should only impact files on disk.

The kernel has a dedicated limit for pipe core dump: https://docs.kernel.org/admin-guide/sysctl/kernel.html#core-pipe-limit

Should we have something similar?

cfreeamd replied:
> I am not sure rlimit.rlim_cur should be used here. It should only impact files on disk.

I'm not sure what you mean here? As I understand it, rlim_cur is the max core size allowed. Is that not what we want?

Looks like the core-pipe-limit is to limit the number of simultaneous processes dumping. Guess we need to decide whether to add that.

palves replied:
We don't need core-pipe-limit on the GPU side.  That exists so that a core filter program can access /proc/pid/foo files without the kernel reusing pid for another process while the core filter program is running.  That problem does not exists on the GPU core side, as the GPU core is generated by the process itself, and of course the process is still around when it itself it dumping a core of itself.

> I'm not sure what you mean here? As I understand it, rlim_cur is the max core size allowed. Is that not what we want?

From "man core":

"However, RLIMIT_CORE will be
          ignored if the system is configured to pipe core dumps to a
          program."

lancesix replied:
argh, I did that one too fast, sorry for the confusion.

So yes, core-pipe-limit is not what we want, and for pipe situations, we probably don't want limits at all.

This however raises one question:  Currently, we use RLIMIT_CORE to know if a core dump should be created at all (and also to know if the runtime should abort on error).  Should we still use this, but ignore the value for the pipe scenario?

I am not sure what the kernel actually does if RLIMIT_CORE is 0 but core_pattern is set to pipe.  Does it completely skip the core, or skip the limit?  I'll try to look into this.

palves replied:
I believe the kernel first checks whether RLIMIT_CORE is zero to determine whether to dump core or not, and that is done even if the pattern is a pipe.  But worth it of a quick test / confirmation.

cfreeamd replied:
Here's what Claude says:

1. Early check: The kernel first checks if rlimit(RLIMIT_CORE) == 0. If so, it aborts the core dump entirely, regardless of whether core_pattern is a file or a pipe.
[ I think we do this currently]

2. Pipe vs file behavior when non-zero:
   * For files: The dump size is limited by RLIMIT_CORE
    * For pipes: The dump size is NOT limited by RLIMIT_CORE (it can be unlimited)

lancesix replied:
As far as I can see, with `ulimit -c 0` I still have a core dump piped to handler.

I have a custom dummy dumper which just copies stdin to some file.

This dummy program:

```c++
#include <sys/resource.h>
#include <cstdlib>
#include <iostream>

int
main (int argc, char **argv)
{
  struct rlimit rlim;
  getrlimit (RLIMIT_CORE, &rlim);
  std::cout << "Curr: " << rlim.rlim_cur << std::endl;
  std::cout << "Max: " << rlim.rlim_max << std::endl;

  std::abort ();
}
```

If I run this locally:

```
$ ulimit -c
0
$ sysctl kernel.core_pattern 
kernel.core_pattern = |/tmp/a.out
$ /a.out baz foo
Curr: 0
Max: 0
Aborted (core dumped)
$ gdb ./a.out -core core.dump
[…]
(gdb) p *argv@3
$3 = {0x7ffe911e100c ".../a.out", 0x7ffe911e1054 "baz", 0x7ffe911e1058 "foo"}
(gdb) p rlim
$4 = {rlim_cur = 0, rlim_max = 0}
```

So it seems the RLIMIT_CORE is ignored by the kernel when dealing with pipe patterns.

This is a bit unfortunate as our runtime completely skips abort if RLIMIT_CORE == 0, so we would have a behaviour different from the system in this case.

lancesix commented on projects/rocr-runtime/runtime/hsa-runtime/libamdhsacode/lnx/amd_core_dump.cpp:
```diff
@@ -304,7 +470,222 @@ struct LoadSegmentBuilder : public SegmentBuilder {
-hsa_status_t build_core_dump(const std::string& filename, const SegmentsInfo& segments, size_t size_limit) {
+// Write core dump to a file descriptor (for pipe handler)
+namespace {
+hsa_status_t write_core_dump_to_fd(int fd, const SegmentsInfo& segments,
```
This feels like quite a bit of duplication.

Could the regular core dump creation be squashed with this one, just doing a `open(2)`, `write_core_dump_to_fd` then `close(2)`?

lancesix commented on projects/rocr-runtime/runtime/hsa-runtime/libamdhsacode/lnx/amd_core_dump.cpp:
```diff
@@ -441,30 +856,71 @@ hsa_status_t build_core_dump(const std::string& filename, const SegmentsInfo& se
+    std::string filename = impl::substitute_core_pattern(pattern);
+
+    if (!use_custom_pattern) {
+      filename += ".gpu";
```
wouldn't that add .gpu after the last fallback case?

```
PREFIX_FILE_NAME + ".%p"
```

I feel this should only be done for pattern coming from the kernel configuration.

lancesix reviewed (commented):
Some comments, I'll try to get back for another pass later.

I only looked for the core changes, not the tests.

cfreeamd committed:
WIP: address code review feedback

cfreeamd commented:
My last commit should address the comments.  As Lancelot mentioned, as it stands now, a gpu core will be suppressed when rlimit is 0 (but not non-zero) when a pipe is used, but this is not the case for CPU cores. Let me know if I should try to make the gpu behavior match this.

I haven't had a chance to fix the rocrtst cases. I have been manually testing with the TestFaultExample. If we need to get this merged soon (I was told today is the last day to make the release; but that it may slip too), I can take out the tests besides TestFaultExample and add those later.

cfreeamd committed:
WIP: gpu core dump rocrtst works; combined

cfreeamd committed:
WIP: remove rocrtst changes for this commit

cfreeamd commented:
I removed the tests from this PR and added them in this one:
https://github.com/ROCm/rocm-systems/pull/1871

I think all the issues have been resolved. Let me know if there's anything else.

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/libamdhsacode/lnx/amd_core_dump.cpp:
```diff
@@ -304,7 +477,12 @@ struct LoadSegmentBuilder : public SegmentBuilder {
+                                          size_t size_limit, bool show_progress) {
+#if defined(__linux__)
   std::unique_ptr<unsigned char[]> copy_buffer(new unsigned char[MAX_BUFFER_SIZE]);
   if (!segments.size()) return HSA_STATUS_SUCCESS;
```
This is not from your change, but could you change the order of these lines here to avoid unnecessary allocation then segments.size() is 0. 

  if (!segments.size()) return HSA_STATUS_SUCCESS;
 std::unique_ptr<unsigned char[]> copy_buffer(new unsigned char[MAX_BUFFER_SIZE]);

cfreeamd commented:
/AzurePipelines run rocm-ci-caller

ScottTodd commented on projects/rocr-runtime/runtime/hsa-runtime/libamdhsacode/lnx/amd_core_dump.cpp:
```diff
@@ -64,8 +70,159 @@ constexpr size_t MAX_BUFFER_SIZE = 4 * 1024 * 1024;
+std::string substitute_core_pattern(const std::string& pattern) {
+  std::string result;
+  pid_t pid = getpid();
+  pid_t tid = gettid();
```
This line of code breaks the build: https://github.com/ROCm/rocm-systems/actions/runs/19448280100/job/55647682668?pr=1732#step:12:8454

```
[amd-llvm] [92/115] Building CXX object runtime/hsa-runtime/CMakeFiles/hsa-runtime64.dir/libamdhsacode/lnx/amd_core_dump.cpp.o
[amd-llvm] FAILED: runtime/hsa-runtime/CMakeFiles/hsa-runtime64.dir/libamdhsacode/lnx/amd_core_dump.cpp.o 
[amd-llvm] /opt/rh/gcc-toolset-12/root/usr/bin/c++ -DHAVE_MEMFD_CREATE -DHSA_DEPRECATED="" -DHSA_EXPORT=1 -DHSA_EXPORT_FINALIZER=1 -DHSA_EXPORT_IMAGES=1 -DHSA_PC_SAMPLING_SUPPORT -DLITTLEENDIAN_CPU=1 -DROCR_BUILD_ID=\"1.18.0-local-build-ae7a88a\" -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -D__linux__ -Dhsa_runtime64_EXPORTS -I/__w/rocm-systems/rocm-systems/projects/rocr-runtime/runtime/hsa-runtime/inc -I/__w/rocm-systems/rocm-systems/projects/rocr-runtime/runtime/hsa-runtime -I/__w/rocm-systems/rocm-systems/projects/rocr-runtime/runtime/hsa-runtime/libamdhsacode -I/__w/rocm-systems/rocm-systems/TheRock/build/compiler/amd-llvm/build/runtimes/rocr-runtime-prefix/src/rocr-runtime-build/runtime/hsa-runtime/core/runtime/trap_handler -I/__w/rocm-systems/rocm-systems/TheRock/build/compiler/amd-llvm/build/runtimes/rocr-runtime-prefix/src/rocr-runtime-build/runtime/hsa-runtime/core/runtime/blit_shaders -I/__w/rocm-systems/rocm-systems/projects/rocr-runtime/libhsakmt/include -isystem /__w/rocm-systems/rocm-systems/TheRock/build/third-party/sysdeps/linux/elfutils/build/dist/lib/rocm_sysdeps/include -isystem /__w/rocm-systems/rocm-systems/TheRock/build/third-party/sysdeps/linux/libdrm/build/stage/lib/rocm_sysdeps/lib/pkgconfig/../../include -isystem /__w/rocm-systems/rocm-systems/TheRock/build/third-party/sysdeps/linux/libdrm/build/stage/lib/rocm_sysdeps/lib/pkgconfig/../../include/libdrm -std=gnu++17 -fPIC -Wall -fPIC -Wl,--unresolved-symbols=ignore-in-shared-libs -fno-strict-aliasing -m64 -msse -msse2 -fexceptions -fno-rtti -fvisibility=hidden -Wno-error=missing-braces -Wno-error=sign-compare -Wno-sign-compare -Wno-write-strings -Wno-conversion-null -fno-math-errno -fno-threadsafe-statics -fmerge-all-constants -fms-extensions -Wno-error=comment -Wno-comment -Wno-error=pointer-arith -Wno-pointer-arith -Wno-error=unused-variable -Wno-error=unused-function -mmwaitx -Wno-deprecated-declarations -Wno-error=maybe-uninitialized -Wno-error=unused-but-set-variable -MD -MT runtime/hsa-runtime/CMakeFiles/hsa-runtime64.dir/libamdhsacode/lnx/amd_core_dump.cpp.o -MF runtime/hsa-runtime/CMakeFiles/hsa-runtime64.dir/libamdhsacode/lnx/amd_core_dump.cpp.o.d -o runtime/hsa-runtime/CMakeFiles/hsa-runtime64.dir/libamdhsacode/lnx/amd_core_dump.cpp.o -c /__w/rocm-systems/rocm-systems/projects/rocr-runtime/runtime/hsa-runtime/libamdhsacode/lnx/amd_core_dump.cpp
[amd-llvm] /__w/rocm-systems/rocm-systems/projects/rocr-runtime/runtime/hsa-runtime/libamdhsacode/lnx/amd_core_dump.cpp: In function ‘std::string rocr::amd::coredump::impl::{anonymous}::substitute_core_pattern(const std::string&)’:
[amd-llvm] /__w/rocm-systems/rocm-systems/projects/rocr-runtime/runtime/hsa-runtime/libamdhsacode/lnx/amd_core_dump.cpp:118:15: error: ‘gettid’ was not declared in this scope; did you mean ‘getuid’?
[amd-llvm]   118 |   pid_t tid = gettid();
[amd-llvm]       |               ^~~~~~
[amd-llvm]       |               getuid
```

cfreeamd replied:
The above must be a rebase error.  Must have left out a header.
I assumed since rocrtst ran and passed, it must have built. I guess with monorepo the tests happen regardless.

For future reference, @ScottTodd can you tell me how you navigated to that link, from this ticket?

ScottTodd replied:
1. Click into the failing github actions job: <img width="935" height="320" alt="image" src="https://github.com/user-attachments/assets/05a911b9-bc54-4ec4-a48e-4fc5f22d9885" />
2. Search within the logs for "error:", "fatal", "failed", or similar to find the issue (often there are multiple log lines, so takes a bit of searching to find the root cause)  <img width="2513" height="804" alt="image" src="https://github.com/user-attachments/assets/f3348869-1666-4f41-b808-892b47ee1aa7" />

cfreeamd replied:
Thanks.  I didn't think ROCr would be in hip-tests, rocprofiler-tests.

ScottTodd replied:
Yeah the naming could use some work:
* https://github.com/ROCm/rocm-systems/blob/fa772be675326381ef7fb6d035dba9c4a600c390/.github/scripts/therock_matrix.py#L19
* https://github.com/ROCm/rocm-systems/blob/fa772be675326381ef7fb6d035dba9c4a600c390/.github/scripts/therock_matrix.py#L24-L27
* https://github.com/ROCm/rocm-systems/blob/fa772be675326381ef7fb6d035dba9c4a600c390/.github/workflows/therock-ci.yml#L77-L78

Looks like it is currently a list of "projects to test" in the job name, which doesn't fully track which projects get built as part of that testing. An exhaustive list would be too long for a job name, so maybe that could instead just mention the component?

cc @geomin12

cfreeamd replied:
That could be helpful, if it's doable.Yeah, there a lot of links now, so any breadcrumbs that allow one to go from their project to a build or run failure would be helpful. OTOH, I understand if there is too much info, that can be confusing as well. 

A confluence page on how to find your build or runtime error could also be helpful. Something you can point to when people like me ask. :)

ScottTodd reviewed (dismissed):
This PR is missing a description. Did you mean to fill out the template, or delete it and write something different?

vbayanag commented:
@ScottTodd @cfreeamd what is the ETA on this PR?

cfreeamd commented:
@vbayanag As soon as it gets an exception (I've tried [here](https://teams.microsoft.com/l/channel/19%3Af50d2d50fd094a8e8065b11a1cfd3496%40thread.tacv2/Gardening%20-%20rocm-systems?groupId=69e7275a-4d49-495d-8bd5-04fb519c8e9c&tenantId=3dd8961f-e488-4e60-8e11-a82d994e183d))  or some of the unrelated CI tests passed.

I'm going to restart the a build hoping the tests (unrelated to this PR) pass.

cfreeamd commented:
/AzurePipelines run rocm-ci-caller

cfreeamd committed:
Merge branch 'develop' into cfreehil-coredump
